project_group_setup.py

```diff
@@ -238,16 +253,24 @@ def main(args):
             project_id_index = identifier_index(identifier_list, "ospoolproject")
             if project_id_index != -1:
                 project_id = str(identifier_list[project_id_index]["Identifier"])
-                is_project = re.compile(OSPOOL_PROJECT_PREFIX_STR + "*").match(project_id) is not None
-            else:
-                is_project = False
+                if re.compile(OSPOOL_PROJECT_PREFIX_STR + "*").match(project_id) is not None:
+                    project_groups.add(gid)
 
             osggid_index = identifier_index(identifier_list, "osggid")
             if osggid_index != -1:
                 highest_osggid = max(highest_osggid, int(identifier_list[osggid_index]["Identifier"]))
-            elif is_project is True:
-                project_name = project_id.replace(OSPOOL_PROJECT_PREFIX_STR, "", 1).lower()
-                projects_to_assign_identifiers.append((gid, project_name,))
+
+            if gid in project_groups:
+                if osggid_index == -1:
+                    project_name = project_id.replace(OSPOOL_PROJECT_PREFIX_STR, "", 1).lower()
+                    project_data = (
+                        gid,
+                        project_name,
+                    )
+                    projects_to_assign_identifiers.append(project_data)
+
+                if not gid in clustered_group_ids:
+                    projects_needing_cluster_groups.add(gid)
```

Comment:
I think this is basically just a `set.difference()`

Reply:
Doesn't set.difference() take another set as it's argument? This could be refactored out of the co_group loop as a difference between project_groups and clustered_group_ids, but we need the co_group loop anyway so I thought this would be better than looping over every project group twice. In this next commit I'll likely shuffle things around in a way that will eliminate the project_groups set altogether.

Reply:
Moved get_projects_to_setup() functionality into a method for each type of setup needed (which use set differences for projects needing cluster groups and unprovisioned projects)